packages/muelu/src/Graph/MatrixTransformation/MueLu_CoalesceDropFactory_def.hpp

```diff
@@ -262,50 +299,160 @@ namespace MueLu {
 
           RCP<Vector> ghostedDiag = MueLu::Utilities<SC,LO,GO,NO>::GetMatrixOverlappedDiagonal(*A);
           const ArrayRCP<const SC> ghostedDiagVals = ghostedDiag->getData(0);
-          const ArrayRCP<bool>     boundaryNodes(A->getNodeNumRows(), false);
+          ArrayRCP<const bool> boundaryNodes = MueLu::Utilities<SC,LO,GO,NO>::DetectDirichletRows(*A, dirichletThreshold);
 
           LO realnnz = 0;
-
           rows[0] = 0;
           for (LO row = 0; row < Teuchos::as<LO>(A->getRowMap()->getNodeNumElements()); ++row) {
             size_t nnz = A->getNumEntriesInLocalRow(row);
             ArrayView<const LO> indices;
             ArrayView<const SC> vals;
             A->getLocalRowView(row, indices, vals);
 
-            //FIXME the current predrop function uses the following
-            //FIXME    if(std::abs(vals[k]) > std::abs(threshold_) || grow == gcid )
-            //FIXME but the threshold doesn't take into account the rows' diagonal entries
-            //FIXME For now, hardwiring the dropping in here
+            if(classicalAlgo == defaultAlgo) {              
+              //FIXME the current predrop function uses the following
+              //FIXME    if(std::abs(vals[k]) > std::abs(threshold_) || grow == gcid )
+              //FIXME but the threshold doesn't take into account the rows' diagonal entries
+              //FIXME For now, hardwiring the dropping in here
+              
+              LO rownnz = 0;
+              for (LO colID = 0; colID < Teuchos::as<LO>(nnz); colID++) {
+                LO col = indices[colID];
+                
+                // we avoid a square root by using squared values
+                typename STS::magnitudeType aiiajj = STS::magnitude(threshold*threshold * ghostedDiagVals[col]*ghostedDiagVals[row]);  // eps^2*|a_ii|*|a_jj|
+                typename STS::magnitudeType aij    = STS::magnitude(vals[colID]*vals[colID]);                                          // |a_ij|^2
+                
+                if (aij > aiiajj || row == col) {
+                  columns[realnnz++] = col;
+                  rownnz++;
+                } else
+                  numDropped++;
+              }
+#if 0
```

Comment:
Is this ifdef disable intentional, and you want to keep this code?

Reply:
fixed